### ISSUE TITLE
Login with Nostr

### ISSUE THREAD

**Issue by lukechilds** [nostr-protocol/nips#154]
I think Nostr makes a good candidate to be used as a very simple DID layer. Having "Login with Nostr" auth on websites solves a lot of problems in a very elegant way, and Nostr's main use case as a social network protocol makes it highly suited to be used as your main identity proving key.

Compare "Login with Nostr" to "Login with Lightning" to see some easy and obvious advantages:

### Remote signer vs local signer

Login with Lightning requires access to remote keys, login with Nostr requires access to local keys ideally stored in a browser extension. Due to the way Lightning works you can only really have one instance. You need all your client devices linked to a single Lightning node, this means most clients will be connecting to the signer remotely. Now if your Lightning node goes down or you lose your connection you also can't auth with any service. This could cause circular dependencies where you lose the connection to your Lightning node so you can't auth with the services you need access to to debug the issue with the connection like your hosting provider or VPN account. You could technically solve this by replicating your LN keys to other client devices only to be used for local auth signing but that introduces other risks.

### Unique identifier vs identity

A Lightning node is not really an identity but a unique identifier. It just tells you the person that auths is the same random person that authed last time, it doesn't tell you who they are. A nostr pubkey is an identity. It tells you who they are, what their name is, what they look like, who they know, how you can pay them, how you can message them.

This is much more useful as an identity layer for an application. The application can show their profile picture, username, send secure cross platform push notifications via NIP-04 encrypted Nostr DMs, etc.

### Consistent identity across services

Lightning pubkeys are sensitive private information and can leak confidential financial information, Nostr pubkeys are safe to share with anyone. LNURL-auth adds extra steps to solve this by creating derived subkeys for identities that are unique to each service you auth with. This does not seem ideal, it seems the default case is that an identity is something that you do want to follow you across all your accounts. Nostr based auth behaves more appropriate in this regard. In the rare case you need to achieve privacy and separation between certain services you can still do that by using use a throwaway Nostr key for those services.

### User relationships across services

Since authing with Nostr shares a real social identity with the service, they can also see your Nostr social graph. This could be useful for connecting you to people you already know on the new service.

## Implementation ideas

This could be done today without any spec changes by just using existing NIP-07 functionality:

```js
const pubkey = async window.nostr.getPublicKey()
const message = {service: 'auth.example.com', challenge}
const signature = async window.nostr.signEvent(message)

if (!isValidSignature(pubkey, signature)) throw new AuthenticationError()

const relays = async window.nostr.getRelays()
const {display_name, picture, nip05} = await fetchNostrMetadata(pubkey, relays)
```

However I think it would make a lot of sense to standardise this as part of the `window.nostr` spec. We already need some standardisation to agree on the challenge object format so it can be safely communicated to the user what they are approving by signing anyway. It would also be nice to abstract this into a single function that all services would depend on instead of everyone having to roll it on their own. It would also be nice for the extension to directly pass over the Nostr user metadata since it probably already has it locally, instead of requiring every single website to directly interface with Nostr and worry about tracking it down from relays.

With a new proposed `window.nostr.authenticate()` method the logic would be 

```js
const {pubkey, signature, display_name, picture, nip05} = await window.nostr.authenticate(nonce)
```

The service string would be set to be the calling origin. You should do some server side validation of `nonce`/`signature` before using this ID for access control. You may also now do additional querying over Nostr directly with `pubkey` for any additional application specific things you want to do, note that this is not required if you just need the basic profile metadata. This makes it very simple for generic websites (not nostr web clients) to integrate with Nostr for auth.

I'm interested in anyones thoughts on this approach and if anyone thinks it would be worth spending more time formalising this into a NIP.

// cc some people who may be interested in this @fiatjaf @jb55 @Cameri @bumi

**Comment by Cameri** [comment]
Concept ACK. Should be trivial to implement in Nos2x. But this only provides identity with an app directly on the browser.
I would want this to work with the proposed NIP called Remote Event Signing.

**Comment by lukechilds** [comment]
@Cameri thanks was not familiar with Remote Event Signing, definitely makes sense to support that too.

Perhaps it could be useful to split this into two different NIPs. A new NIP to define the auth note schema (which both browser extensions and remote signers would know how to interpret). And separately an extension to NIP-07 to add the `window.nostr.authenticate()` method.

**Comment by lylepratt** [comment]
agree that login with nostr offers a lot of wins!

we implemented "Login With Nostr" on https://vida.page via a simple 2-Factor DM flow. it's a UX users are familiar with and works on web+mobile. 

basically the user gives us a pubkey or a nip-05. if nip-05 we check for relays and auto fill, if just pubkey we also ask them for a relay. then we send them a pin via encrypted DM and ask them to input the pin. if successful, we will their public details via the same relay and provision their account.

**Comment by lylepratt** [comment]
I have a little library I made with helper functions based on @fiatjaf's nostr-tools lib to make the flow easy. just need to clean it up and will publish if anyone is interested but none of the flow is complex.

**Comment by lylepratt** [comment]
one way to clean this up from a UX perspective in clients would be to agree on a format for the DM that is sent so that clients could automatically pop-up an alert with the pin (similar to Apple's 2-factor auth) if it receives a specifically formatted message. 

So, optimally, service sends the DM, user opens their client and is immediately presented the pin.

**Comment by lukechilds** [comment]
@lylepratt that's a very cool approach, good point re browser APIs not working on mobile.

One issue you could have with a DM based approach is going through a relay. I really like how my above approach does not rely on Nostr for networking, only for signing, it makes auth much more reliable.

Since Nostr only kinda half guarantees that stuff ends up where it should, it seems not ideal to rely on going through relays for something critical like auth. For example what if I have my client configured so low PoW DMs are ignored. I wouldn't see your DM if you don't add enough PoW. Even if you add PoW what if my client is configured to ~$1 PoW per DM, can you justify paying that to log a user in? Or what if my main relay goes down and you can't find me, I can't authenticate with any services until it comes back online.

I wonder if we could use the same approach you explained above but using data URIs to jump between apps instead of DMs. 

The web app (or even a native app) could encode a payload like:

```
{
  service: 'auth.vida.page',
  challenge: 'ba65d3b01e6bc841050a3525133d38d2',
  callback: 'https://auth.vida.page?nauth_response=<token>'
}
```

into a Nostr URI like:

```
nostr:nauth1blahblahblahblah
```

Which gets handled by your native Nostr client app and decoded so no actual networking is required.

The client app then knows to prompt the user if they want to log in to `auth.vida.page`, signs the challenge, encodes the signature along with user metadata into a token and opens `https://auth.vida.page?nauth_response=blah` which can then be caught and handled by the original website (or app via URI handler).

The original service can now decode the token, verify the challenge/response sig server side and has all the users metadata.

**Comment by Cameri** [comment]
> @lylepratt that's a very cool approach, good point re browser APIs not working on mobile.
> 
> 
> 
> One issue you could have with a DM based approach is going through a relay. I really like how my above approach does not rely on Nostr for networking, only for signing, it makes auth much more reliable.
> 
> 
> 
> Since Nostr only kinda half guarantees that stuff ends up where it should, it seems not ideal to rely on going through relays for something critical like auth. For example what if I have my client configured so low PoW DMs are ignored. I wouldn't see your DM if you don't add enough PoW. Even if you add PoW what if my client is configured to ~$1 PoW per DM, can you justify paying that to log a user in? Or what if my main relay goes down and you can't find me, I can't authenticate with any services until it comes back online.
> 
> 
> 
> I wonder if we could use the same approach you explained above but using URI handlers to jump between apps instead of Nostr DMs. 
> 
> 
> 
> The web app (or even a native app) could encode a payload like:
> 
> 
> 
> ```
> 
> {
> 
>   service: 'auth.vida.page',
> 
>   challenge: 'ba65d3b01e6bc841050a3525133d38d2',
> 
>   callback: 'https://auth.vida.page?nauth_response=<token>'
> 
> }
> 
> ```
> 
> 
> 
> into a Nostr URI like:
> 
> 
> 
> ```
> 
> nostr:nauth1blahblahblahblah
> 
> ```
> 
> 
> 
> Which gets handled by your native Nostr client app and decoded so no actual networking is required.
> 
> 
> 
> The client app then knows to prompt the user if they want to log in to `auth.vida.page`, signs the challenge, encodes the signature along with user metadata into a token and opens `https://auth.vida.page?nauth_response=blah` which can then be caught and handled by the original website (or app via URI handler).
> 
> 
> 
> The original service can now decode the token, verify the challenge/response sig server side and has all the users metadata.
> 
> 
> 
> Thoughts?

I prefer this approach much more.

**Comment by lukechilds** [comment]
>I prefer this approach much more.

You mean prefer it to DMs or prefer it to `window.nostr.authenticate()`?

**Comment by Cameri** [comment]
> >I prefer this approach much more.
> 
> 
> 
> You mean prefer it to DMs or prefer it to `window.nostr.authenticate()`?

I prefer URI handlers as opposed to DMs. I think we still need to support window.nostr.authenticate() if available, otherwise fall back to URI handler. Sorry for the confusion.

**Comment by lukechilds** [comment]
Yep I agree, `window.nostr.authenticate()` is nice to have since it's detectable and can be automated instead of requiring a URI button press.

**Comment by lylepratt** [comment]
I like the encoded payload idea. it's very similar to this idea I tried to explain here: https://twitter.com/lylepratt/status/1610829070928908288?s=46&t=HyWgoORCb6b2g5WZD7RRug

"This is an idea for a NIP that would make it easy for external apps, services, and APIs to automatically integrate with Nostr clients.

- Sign a message containing a timestamp within X seconds (or minutes) of “now”.

- Client appends the signed message as a URL param (or via POST) in supported links that get presented to users.

- The Service sees the param, fetches and provisions profile info if needed, and grants access to the service to the user.

I’m going to try to do something like this on Vida. For example, it would let us make it easy for Nostr clients to link a user directly into an audio/video spaces room (or even pull external streams into the app directly via API). Lots of possibilities for other stuff too!"

**Comment by Cameri** [comment]
Relevant perhaps: https://github.com/nostr-protocol/nips/blob/329cd8d8a19fd1dcb12df7eb7a5d640b2a5b6bf8/19.md#shareable-identifiers-with-extra-metadata

**Comment by fiatjaf** [comment]
Why not just request the user to sign an event? I fail to understand why this added complexity is accomplishing.

**Comment by lukechilds** [comment]
@Cameri niceee, that looks perfect for this!

**Comment by lukechilds** [comment]
>Why not just request the user to sign an event? I fail to understand why this added complexity is accomplishing.

Could that not be dangerous?

If apps just send out generic events for auth then what stops a malicious site MITMing a user and requesting them to sign a challenge for a different site? It may not be clear to the user what they are signing from the payload. If some kind of auth object is standardised then it can show a very clear message explain that this is logging the user in to `example.com` and block attempts to sign this message from origins other than example.com.

**Comment by fiatjaf** [comment]
Of course if we make a super safe multi-step protocol that can be pretty safe, but I am thinking about the fallback case, i.e. do you need a new implementation from _every client_ for this to work? Or can clients that not implement this but that can still sign arbitrary events still work, even with worse UX?

**Comment by lukechilds** [comment]
Also to TLDR:

>Why not just request the user to sign an event?

That basically is all that's happening here but in a standardised way so:
- Any Nostr signing app/extension/website can sign for any other app/website without having to manually copy/paste JSON strings around
- It can be done in a secure way that prevents MITM attacks
- It explains to users what they are doing without expecting them to understand the meaning behind the JSON string

**Comment by lukechilds** [comment]
>do you need a new implementation from every client for this to work? Or can clients that not implement this but that can still sign arbitrary events still work, even with worse UX?

Yes if you can sign a note you can authenticate with this protocol. But if your signer didn't understand this protocol you'd need to manually copy/paste JSON between the apps instead of clicking buttons, and you wouldn't get a warning if `evil.com` asks you to sign a challenge for `coinbase.com`.

**Comment by lukechilds** [comment]
To make this more generic instead of `nostr:nauth1blahblahblah` maybe it could be `nostr:nsign1blahblahblah` which when clicked asks whichever app has registered a URI handler for `nostr` to sign the encoded note.

Then Nostr clients can optionally detect an app authentication note and parse the metadata and follow this protocol. Dumb clients would just blindly sign which could be dangerous if the user doesn't understand what the JSON means and you'd have to paste the result back over in the initial app.

**Comment by fiatjaf** [comment]
Hmm, yes, thinking better about this I think it makes sense to have the `nostr:...` something and the "fallback" case I had in mind isn't really possible at all today so we're not losing anything by going with a more manual approach. The browser extensions can just intercept clicks on these `nauth` links and handle them natively. I like this very much.

**Comment by fiatjaf** [comment]
> `display_name, picture, nip05` nope, signer extensions should not have to connect to relays. this could be added if the signing extension has a config option for what metadata it should offer to websites but otherwise apps should query kind 0 themselves.

Yes, but I would say this shouldn't even be optional, it should be forbidden. The apps should just fetch metadata directly if they want. What the signing app should do is give a relay URL to the website from where they can fetch the metadata.

**Comment by fiatjaf** [comment]
Shouldn't this thing be somehow merged with the "Nostr Connect" proposal and also with the Relay Authentication" proposal? These are all separate things, but they may share something in common.

**Comment by lukechilds** [comment]
@Semisol @fiatjaf good points re not returning metadata. I was thinking it made it super simple for sites to integrate without having to do too much Nsotr stuff, but agree it pushes too much complexity over to the signer.

>Shouldn't this thing be somehow merged with the "Nostr Connect" proposal and also with the Relay Authentication" proposal? These are all separate things, but they may share something in common.

I'm not familiar with those but will look into them.

**Comment by fiatjaf** [comment]
By the way, `lnurl-auth` has gained a lot of love in the past years and many users and websites seem to love the idea, however the protocol has a bunch of annoying flaws and UX issues in my opinion. Me and @Dolu89 and some others I don't remember were discussing replacing it with a better protocol, and in my mind that could be a protocol generic enough to support keys from other sources other than Bitcoin wallets -- i.e. if you can produce a bip340 signature you can have an identity and login to a website. Shouldn't we make this generic enough to allow this to be that unified protocol? Could be just by using a generic url scheme like `keyauth:...` and then producing a normal signature over a challenge, not a Nostr event specifically.

**Comment by Cameri** [comment]
> By the way, `lnurl-auth` has gained a lot of love in the past years and many users and websites seem to love the idea, however the protocol has a bunch of annoying flaws and UX issues in my opinion. Me and @Dolu89 and some others I don't remember were discussing replacing it with a better protocol, and in my mind that could be a protocol generic enough to support keys from other sources other than Bitcoin wallets -- i.e. if you can produce a bip340 signature you can have an identity and login to a website.
> 
> Shouldn't we make this generic enough to allow this to be that unified protocol? Could be just by using a generic url scheme like `keyauth:...` and then producing a normal signature over a challenge, not a Nostr event specifically.

I would love this! But should we support Nostr first before we go nuts like with DID methods?

**Comment by fiatjaf** [comment]
The signature must contain the domain name otherwise it can be phished. I would do this change:

```diff
-client should sign sha256 of the JSON ["keyauth", <challenge string>, <unix epoch seconds at time of signing>] encoded without unnecessary whitespace, and should send a POST to https://example.com/endpoint?challenge=challenge (along with any other query parameters in the keyauth:// URI) with the following added query parameters
+client should sign sha256 of the JSON ["keyauth", "keyauth://example.com/endpoint?challenge=challenge", <unix epoch seconds at time of signing>] encoded without unnecessary whitespace, and should send a POST to https://example.com/endpoint?challenge=challenge (along with any other query parameters in the keyauth:// URI) with the following added query parameters
```

**Comment by fiatjaf** [comment]
> But should we support Nostr first before we go nuts trying to support everything like with DID methods?

I think we should support just bip340 signatures. That includes Bitcoin wallets, Nostr clients and other things that may show up later and decide to reuse our network-effect. They will be forced to use bip340 signatures and nothing more than that.

**Comment by lukechilds** [comment]
Just my 2c, but I agree with this:

>But should we support Nostr first before we go nuts trying to support everything like with DID methods?

It seems like extra URI handlers and more stuff to worry about that likely nobody is gonna use right now. `nsign` would also be useful to any app that wants anything signed, not just authenticating services. Also is BIP340 used for much outside of Nostr and Bitcoin? If a hard requirement is BIP340 then it seems like you're pretty much limited to Nostr or Bitcoin, and for the reasons mentioned in the initial post I think Nostr makes way more sense than Bitcoin for this.

If you for some reason want to use your Bitcoin keys to auth you can still do that with this protocol via BIP39 > NIP-06 > this. I get that's not as clean but again I don't think it makes much sense to support this. A huge benefit of this protocol is that it's an actual identity, not just a unique identifier.

**Comment by lylepratt** [comment]
> simple keyauth proposal:
> 
> - clients have a public/private keypair
> 
> - the URI scheme is `keyauth://example.com/endpoint?challenge=challenge`
> 
> - client should sign sha256 of the JSON `["keyauth", <challenge string>, <unix epoch seconds at time of signing>]` encoded without unnecessary whitespace, and should send a POST to `https://example.com/endpoint?challenge=challenge` (along with any other query parameters in the `keyauth://` URI) with the following added query parameters
> 
>   - `pk`: pubkey encoded as hex
> 
>   - `signed_at`: the unix epoch that was used for signing
> 
>   - `sig`: the signature as hex

I like this. very similar to what I was thinking. The service could even declare its support for this via a .well-known which could enable direct linking into apps via GETs with pre-fetched signed challenges included via a query param.

**Comment by fiatjaf** [comment]
> Also is BIP340 used for much outside of Nostr and Bitcoin?

Before Nostr it was only used for Bitcoin, so no one tried to make a thing that was generic enough. If they had Nostr could just use that. We don't know what will be created after Nostr.

> Nostr makes way more sense than Bitcoin for this

Sure, but people are using lnurl-auth successfully in a lot of places right now. They're not using Bitcoin keys per se, but they're making the signatures from Bitcoin and Lightning wallets. And also from custodial providers that generate virtual hosted keys for accounts and allow them to sign in. That's proof that demand for this exists outside of Nostr.

---

On the other hand I don't like generic stuff for the sake of genericness. I am against that too. I am just considering that this may be an actual justifiable situation in which genericness is desirable.

**Comment by lukechilds** [comment]
How do you see the social side of this more generic protocol working?

So once you've verified the sig you figure out what the pubkey is used for and do lookups based on that? If it's a Nostr key you query their relay. If it's a Bitcoin key you fill out some generic user info or potentially resolve it on some hypothetical `bitcoinid.com/<address>` server. If it's some other future BIP340 based social thing you fetch the user meta from that. Is that how you see it working?

**Comment by fiatjaf** [comment]
Hmm, no, I hadn't thought about any of that.

Now that you said that I am convinced we should do this as just a Nostr thing, and if other protocols want to use it they can always generate a Nostr identity with optional metadata.

**Comment by lukechilds** [comment]
Yeah this is inline with my thoughts too. Also consider, it's not like Nostr is some super heavy protocol, and Nostr libs are already available in many languages. So potentially having a separate generic protocol for this actually makes it harder for other things to integrate with because they may not have an existing lib for this generic protocol in their lang. I know it would be super minimal formatting convention stuff over BIP340 but it still wouldn't be as minimal as:

````
import nostr

nostr.nip_x(url, challenge, callback)
...
nostr.verify(pubkey, callbackResult)
```

**Comment by lukechilds** [comment]
So if we're going the Nostr only route then we need to do two things:

1. A URI handler/entity combo to pass the event from the app to the signer that can work with any combo of native/web/extension apps.
1. A scheme for a special event type that is understood to be signed and posted to a callback to log the user in to a service.

### URI Handler / Entity

A new `nostrsign` uri which expects to be passed an event to sign. We want a new uri because the app that signs may be different to the app that browses Nostr which may already have registered the `nostr` handler.

```
nostrsign:nevent1<base58 encoded note>
```

### Auth event

The event should contain a service id like `auth.example.com`. This is clearly displayed to the user as the service being authenticated. A challenge, which can be anything but should be a high entropy random hex string. And also callback URI like `https://auth.example.com/?token=<token>`. Some magic value like `<token>` will act as a template to be replaced with the resulting signed note.

I think the best Nostr way to do this is with a new event kind with a JSON string value?

So maybe something like kind 8 `authentication_challenge` set to a stringified JSON object of:

```js
{
  service: 'auth.example.com',
  challenge: 'ba65d3b01e6bc841050a3525133d38d2',
  callback: 'https://auth.example.com?token=<token>'
}
```

Does that make sense?

### Usage

So to clarify the flow incase it is unclear.

Lets imagine I'm using a generic 3rd party service, say github.com, and I want to authenticate via the Damus app on iOS.

It's actually fairly simple and there's really only 3 main steps that happen but i'll break it down really pedantically to make everything very clear.

- GitHub generates a random challenge nonce `ba65d3b01e6bc841050a3525133d38d2` serverside
- GitHub generates `authentication_challenge` Nostr event with the payload:
```
{
  service: 'auth.github.com',
  challenge: 'ba65d3b01e6bc841050a3525133d38d2',
  callback: 'https://auth.github.com?token=<token>'
}
```
- GitHub encodes event into `nostrsign` URI like: `nostrsign:nevent1<base58 encoded note>`
- GitHub shows a "Login with Nostr" button with it's href set to the above URI
- I click the button in Safari on my iPhone
- Damus has registered the `nostrsign` URI handler so gets passed the URI
- Damus decodes the note, recognises it is an `authentication_challenge` event and asks me if I want to log in to `auth.github.com`
- I manually click to approve
- Damus adds my relays to the event
- Damus signs the event
- Damus adds the signed event to the callback URI template like `https://auth.github.com?token=<signed-event>`
- Damus opens the callback URI
- I now get shown Safari loading the callback uri
- GitHub grabs the signed event from the URI
- GitHub verifies they recognise the challenge nonce
- GitHub verifies the signature of the event
- GitHub queries my Nostr relay and fetches my metadata
- GitHub creates a user entry in their DB for me with my Nostr pubkey as my unique id and maps Nostr display name, profile pic, etc to their metadata
- I am now logged in on GitHub.com with a rich user profile

Some notes/thoughts:

Since URIs are used both for passing the challenge to the signer and passing the signed response back to the original app, this protocol will work with native apps, browser extensions and web apps since they can all register URI handlers. Any combination of the two can communicate.

- A web app can get authed by a native signer.
- A native app can get authed by a browser extension signer.
- A native app can get authed by a web signer.

Re the `Damus adds my relays to the event` is there an existing way to attach relays to a note? It's looks like potentially "e" tags can do this if they can self reference? If not we could just add them to the `authentication_challenge` JSON string before signing.

A web extension signer can do an extra sanity check by checking if the origin webpage URL matches the `service` and refusing to sign if not. I'm not sure if native signers activated by URI handlers are able to see any referrer information about who called them. If they can, potentially we can lock something down there, if not all we can do is make the `service` value very clear to the user and explain that the action logs them in to that service.

**Comment by lukechilds** [comment]
Thinking further about this. Maybe we can simplify by dropping `service` and instead just do:

```js
{
  challenge: 'ba65d3b01e6bc841050a3525133d38d2',
  callback: 'https://auth.example.com?token=<token>'
}
```

The service to show to the user is the hostname of the callback. This is also safer as you you're showing the user the hostname that actually receives the token.

**Comment by lukechilds** [comment]
Just a quick summary from https://github.com/nostr-protocol/nips/pull/153#issuecomment-1374894006 that might be useful to get a quick overview of this protocol if https://github.com/nostr-protocol/nips/issues/154#issuecomment-1374681051 is too verbose:

Login with Nostr defines a `nostrsign:nevent1<base58 encoded event> to ask whichever app has registered a `nostrsign` handler to sign the provided event. If the event is of the new `authentication_challenge` kind then it will show some anti phishing UI elements to explain you're logging in to an app and send the signed result to the callback URI specified within the note.

**Comment by lukechilds** [comment]
@Semisol 

>unnecessary complexity with base58 encoding, use base64url in my opinion.
>also, binary encoding should be looked into to not have extremely long sign URLs

Ahh I misunderstood NIP-19. I thought it encoded an entire event object. Not just an event ID + relay. Also very good point re URL length. Is binary Nostr event serialisation standardised anywhere? That should solve both the issues you raised.

For the URI:

```js
`nostrsign:${base64(binarySerialise(event))}`
```

For the callback:

```js
`https://auth.example.com?token=${base64(binarySerialise(signedEvent))}`
```

Here's what a dummy `authentication_challenge` event might look like. Simple base64 encoded text serialisation gets us down to ~336 chars which isn't too bad. Agree would be great to get it lower with binary serialisation. I don't see a spec for that, should I roll it myself?


```js
> event
{
  id: '0000000000000000000000000000000000000000000000000000000000000000',
  kind: 8,
  pubkey: '0000000000000000000000000000000000000000000000000000000000000000',
  created_at: 1673204003688,
  content: '{"challenge":"ba65d3b01e6bc841050a3525133d38d2","callback":"https://auth.example.com?token=<token>","relays":["wss://foo.bar","wss://damus.io"]}',
  sig: '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'
}
> serializeEvent(event)
'[0,"0000000000000000000000000000000000000000000000000000000000000000",1673204003688,8,null,"{\\"challenge\\":\\"ba65d3b01e6bc841050a3525133d38d2\\",\\"callback\\":\\"https://auth.example.com?token=<token>\\",\\"relays\\":[\\"wss://foo.bar\\",\\"wss://damus.io\\"]}"]'
> Buffer.from(serializeEvent(event)).toString('base64')
'WzAsIjAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAiLDE2NzMyMDQwMDM2ODgsOCxudWxsLCJ7XCJjaGFsbGVuZ2VcIjpcImJhNjVkM2IwMWU2YmM4NDEwNTBhMzUyNTEzM2QzOGQyXCIsXCJjYWxsYmFja1wiOlwiaHR0cHM6Ly9hdXRoLmV4YW1wbGUuY29tP3Rva2VuPTx0b2tlbj5cIixcInJlbGF5c1wiOltcIndzczovL2Zvby5iYXJcIixcIndzczovL2RhbXVzLmlvXCJdfSJd'
> Buffer.from(serializeEvent(event)).toString('base64').length
336
```

**Comment by trbouma** [comment]
This is a simple login for LNURLw that I implemented a few months back.

The random challenge is also the corresponding  the token grant which then is used to set the cookie. The page polls to see if the challenge is signed and the corresponding  the token grant is issued. Since it is random, only the signer should know what token grant to look for. In theory, the user could sign anything, but it’s safer to issue something unique for the user to sign and knows what to look for. 

https://fastlogin.io/zlogin

**Comment by HamishMacEwan** [comment]
> A nostr pubkey is an identity. It tells you who they are, what their name is, what they look like, who they know, how you can pay them, how you can message them.

"what they look like⁉" 🤣

It isn't and it doesn't.  It's akin to a self-signed certificate and those aren't treated with much respect.

> Due to the way Lightning works you can only really have one instance. 

Due to the way LNbits accounts work, you can have as many as you wish.

**Comment by lukechilds** [comment]
@HamishMacEwan

>"what they look like⁉" 🤣
>
>It isn't and it doesn't. It's akin to a self-signed certificate and those aren't treated with much respect.

Most users publish a kind 0 `set_metadata` event associated with their pubkey which includes things like their name, and a picture. Self-signed certificates generally don't have any kind of reputation attached to them. Nostr pubkeys can gain reputation via PoW/NIP-05/social graph.

<img src="https://user-images.githubusercontent.com/2123375/213008258-1a5b76e7-e079-4206-8ed0-a914cfef1234.jpg" width="400" />

**Comment by lukechilds** [comment]
@trbouma thanks for the explanation, I appreciate it.

>if they have access to the authentication request channel, they can steal the grant token, but it is assumed that there is already a secure (https) session between authenticating client and the service.

To be clear the attack vector is:
- Attacker sets up a webpage at evil.com that looks like coolwebsite.com.
- Attacker tricks a user into visiting evil.com with the user thinking it is coolwebsite.com
- user goes to login via LNURL
- evil.com makes a request to coolwebsite.com to login via LNURL, shows the challenge URL to the user on evil.com
- evil.com keeps polling the endpoint on coolwebsite.com to check if the challenge is signed
- user scans the QR, thinking they are logging in to coolwebsite.com
- users app signs and submits the sig to the coolwebsite.com callback
- coolwebsite.com validates everything from users pubkey
- evil.com hits poll endpoint again, does coolwebsite.com now notice the challenge has been signed and return an auth token to evil.com?

If so the poll logic has introduced a phishing attack vector because the server is assuming that whoever is polling is the same person who is signing, but that isn't guaranteed to be the case.

That isn't the case with OAuth because when you send the user off to a third party to be verified the third party then sends them directly back to the original site and they get authed. There's no polling some identifier that allows you to hijack the session once something has been done.

I'm not saying this is a major issue, I'm just trying to understand. There are similar attacks you can do against OAuth where you phish the OAuth provider screen, grab the users auth credentials, then forward them on to the real Oauth provider to verify. However that requires the user to misread the Oauth providers URL, something users a slightly better trained to do. The attack vector described above works if the user misreads any of the client website URLs which is slightly more likely.

One solution to solve this issue is to return the token to the signer so only the signing device is authed. That prevents phishing, but has the (quite significant) inconvenience that if you start the login flow on desktop and sign via a mobile app, you then get authed in a mobile browser, not your desktop.

Or maybe I'm still misunderstanding your implementation?

**Comment by trbouma** [comment]
The wallet has to make independent callback - if the callback is evil.com that is what will be presented in the wallet and the user will note that it is different than coolsite.com. As well, the resulting access token is signed and encrpyted to only work with coolsite.com so it wont won't work with evil.com. 

Anyway, it's been a few months since I implemented it, and I consulted this doc pretty closely to make sure I had the flows right. https://datatracker.ietf.org/doc/draft-ietf-gnap-core-protocol/

**Comment by lukechilds** [comment]
Right, but in the above example the wallet will make a callback to coolwebsite.com, the issue is that evil.com is polling coolwebsite.com waiting for that challenge to be posted and when it does, coolwebsite.com gives a session to evil.com mistakenly thinking it is giving a session to a user.

I have just verified and can confirm fastlogin.io is vulnerable to this attack

If I go to login I'm presented with a QR to scan and the page polls `https://fastlogin.io/token/<identifier>` until the QR has been authenticated at which point the poll endpoint returns a 301 redirect with a `Set-Cookie: fastlogin_jwt=<auth-token>` header.

If I register fastlogin.xyz, show the same UI as fastlogin.com, have my server dynamically grab auth QRs from your site and show it on mine, and keep polling the endpoint on my server, if I then trick a user to login to my phishing site, your server will return the auth cookie to my server, not the user, and my server will now be authenticated as the user.

**Comment by Dolu89** [comment]
Are every websites using LNURL-Auth not vulnerable to this? I think they are

**Comment by lylepratt** [comment]
- Assume a Nostr client wants to link a user to another website (or embed a website) so that when they get there, they land on the website already logged into a rich profile.
- The Nostr client queries a "well-known" endpoint to see if the web site supports it. If so, it encrypts a message (using the service's pubkey found in the well-known) containing a timestamp of "now" and appends it as a query param (or post).
- The website sees this special param, decodes the message and checks the timestamp. If within X seconds, it lets the user in, fetches their nostr profile details, stubs out a profile, and the user lands in a rich experience (perhaps like a nostr compatible twitter spaces widget that could even be loaded in an iframe).

This is a slightly different scenario than a user landing on a website and clicking a "Login with Nostr" button, potentially providing a gateway for mini-web apps that could be embedded into Nostr clients. 

Is something like this possible with what you envision so far @lukechilds? Perhaps there are inherent problems with my example that I haven't thought through, but perhaps its still possible with your envisioned NIP.

**Comment by Dolu89** [comment]
@lukechilds With you current process (see https://github.com/nostr-protocol/nips/issues/154#issuecomment-1374681051), we can't have a key in a dedicated app, or in Damus and using it to login on a computer by flashing a QR code.

I think it's important. I would like to login in many places without installing browser extension like Alby everywhere and copy/paste my private key inside it.

**Comment by Dolu89** [comment]
Also, while discussing on [twitter](https://twitter.com/Dolu_Web/status/1623743354201595912) about it, what do you think if we extract your protocol from Nostr and make the step "Damus adds my relays to the event" optional?

This protocol could be called something like "Keyauth V2" or something else, it could works without Nostr by default, so no rich profile, you are just identified by your public key.
Then, other protocols could add some optional params, like Nostr with relays params.

`Signed event`
```json
{
  "challenge": "ba65d3b01e6bc841050a3525133d38d2",
  "callback": "https://auth.example.com?token=<signed-token>",

  // Optional params to fetch profile from somewhere
  "nostr": {
    "relays": ["wss://relay1.com", "wss://relay2.com"]
  },
  "futureNewDecentralizedProtocol": {
    "servers": ["https://server1.com", "https://server2.com"]
  },
}
```

By using this, any compatible https://bips.xyz/39 and https://bips.xyz/340 key holders could be used

**Comment by fiatjaf** [comment]
I like @Dolu89's proposal.

**Comment by PeerRich** [comment]
once we found consent I'd love to promote "Sign in with Nostr" at https://authjs.dev/, I know the team behind it. It's a widely popular auth library for any javascript app

**Comment by ekzyis** [comment]
@PeerRich sorry to hijack this discussion but if you know someone from `next-auth`, maybe you can promote the following ticket for LNURL-auth: https://github.com/nextauthjs/next-auth/issues/7872

**Comment by PeerRich** [comment]
@ekzyis will do!

**Comment by bijeebuss** [comment]
> @lukechilds With you current process (see [#154 (comment)](https://github.com/nostr-protocol/nips/issues/154#issuecomment-1374681051)), we can't have a key in a dedicated app, or in Damus and using it to login on a computer by flashing a QR code.
> 
> I think it's important. I would like to login in many places without installing browser extension like Alby everywhere and copy/paste my private key inside it.

I think you could still sign into desktop with your phone using this method but the login UI would have to have an option to either open the url directly or display it as a QR code.

**Comment by damascene** [comment]
Seems like a nice idea. Is there any progress on this?
